File: scripts/verify-apkg-import.mjs
import { spawnSync } from "node:child_process";
import { access, mkdtemp, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join, resolve } from "node:path";
import { buildApkg } from "../src/anki/apkg.ts";
import { createDraft } from "../tests/fixtures.js";

const ankiRoot = join(
  process.env.LOCALAPPDATA ?? "",
  "AnkiProgramFiles",
);
const python = join(ankiRoot, ".venv", "Scripts", "python.exe");
await access(python);

const root = await mkdtemp(join(tmpdir(), "yt2anki-apkg-import-"));
const firstPath = join(root, "first.apkg");
const repeatPath = join(root, "repeat.apkg");
const collectionPath = join(root, "collection.anki2");
const wasmPath = resolve("node_modules/sql.js/dist/sql-wasm.wasm");
globalThis.chrome = {
  runtime: {
    getURL: () => wasmPath,
  },
};

const firstDraft = createDraft({
  target: "packaged original",
  title: "Import fixture",
  translation: "Hello.",
});
const first = await buildApkg({
  deckName: "yt2anki::Import fixture",
  draft: firstDraft,
  segments: firstDraft.segments,
});
await writeFile(firstPath, first);

const repeatDraft = structuredClone(firstDraft);
repeatDraft.segments[0].target = "packaged replacement";
repeatDraft.segments.push({
  ...repeatDraft.segments[0],
  endMs: 6_250,
  identity: "v1_BBCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcde",
  sourceEndMs: 6_000,
  sourceStartMs: 4_000,
  startMs: 3_750,
  target: "再见。",
  translation: "Goodbye.",
});
const repeat = await buildApkg({
  deckName: "yt2anki::Import fixture",
  draft: repeatDraft,
  segments: repeatDraft.segments,
});
await writeFile(repeatPath, repeat);

try {
  const result = spawnSync(
    python,
    [
      resolve("scripts/verify_apkg_import.py"),
      collectionPath,
      firstPath,
      repeatPath,
      firstDraft.segments[0].identity,
    ],
    { encoding: "utf8" },
  );
  if (result.status !== 0) {
    throw new Error(
      [
        "Disposable Anki package import failed.",
        result.stdout.trim(),
        result.stderr.trim(),
      ]
        .filter(Boolean)
        .join("\n"),
    );
  }
  const report = JSON.parse(result.stdout);
  if (
    report.first.new !== 1 ||
    report.repeat.new !== 1 ||
    report.repeat.duplicate !== 1 ||
    report.noteCount !== 2 ||
    report.cardCount !== 2 ||
    report.existingTarget !== "local edit" ||
    !report.schedulingPreserved ||
    report.noteTypeCount !== 1 ||
    report.deckCount !== 1 ||
    report.rootDeckCount !== 1 ||
    !report.cardsInVideoDeck
  ) {
    throw new Error(`Unexpected Anki import report: ${result.stdout}`);
  }
  console.log(
    "Verified final .apkg repeat import in disposable Anki 25.09.5.",
  );
} finally {
  const resolvedRoot = resolve(root);
  const resolvedTemp = resolve(tmpdir());
  if (!resolvedRoot.startsWith(`${resolvedTemp}\\`)) {
    throw new Error("Refusing to remove a non-temporary integration path.");
  }
  await rm(resolvedRoot, { force: true, recursive: true });
}

File: scripts/verify_apkg_import.py
import json
import sys
from pathlib import Path

from anki.collection import Collection, ImportAnkiPackageRequest


collection_path, first_path, repeat_path, existing_identity = map(
    Path, sys.argv[1:]
)
collection = Collection(str(collection_path))

try:
    options = collection._backend.get_import_anki_package_presets()
    first = collection.import_anki_package(
        ImportAnkiPackageRequest(
            package_path=str(first_path),
            options=options,
        )
    )
    existing_ids = collection.find_notes(
        f'SegmentIdentity:"{existing_identity}"'
    )
    assert len(existing_ids) == 1
    existing = collection.get_note(existing_ids[0])
    existing["Target"] = "local edit"
    collection.update_note(existing)

    card = collection.get_card(existing.card_ids()[0])
    card.due = 4242
    card.ivl = 17
    card.queue = 2
    card.type = 2
    collection.update_card(card)
    scheduling = (card.due, card.ivl, card.queue, card.type)

    repeat = collection.import_anki_package(
        ImportAnkiPackageRequest(
            package_path=str(repeat_path),
            options=options,
        )
    )
    existing = collection.get_note(existing_ids[0])
    card = collection.get_card(existing.card_ids()[0])
    note_type_name = "yt2anki Listening v1"
    deck_name = "yt2anki::Import fixture"

    report = {
        "first": {
            "duplicate": len(first.log.duplicate),
            "new": len(first.log.new),
            "updated": len(first.log.updated),
        },
        "repeat": {
            "duplicate": len(repeat.log.duplicate),
            "new": len(repeat.log.new),
            "updated": len(repeat.log.updated),
        },
        "existingTarget": existing["Target"],
        "schedulingPreserved": (
            card.due,
            card.ivl,
            card.queue,
            card.type,
        )
        == scheduling,
        "noteCount": len(
            note_ids := collection.find_notes(f'note:"{note_type_name}"')
        ),
        "cardCount": sum(
            len(collection.get_note(note_id).card_ids())
            for note_id in note_ids
        ),
        "noteTypeCount": sum(
            item.name == note_type_name
            for item in collection.models.all_names_and_ids()
        ),
        "deckCount": sum(
            item.name == deck_name
            for item in collection.decks.all_names_and_ids()
        ),
        "rootDeckCount": sum(
            item.name == "yt2anki"
            for item in collection.decks.all_names_and_ids()
        ),
        "cardsInVideoDeck": all(
            collection.decks.name(collection.get_card(card_id).did)
            == deck_name
            for card_id in collection.find_cards(
                f'note:"{note_type_name}"'
            )
        ),
    }
    print(json.dumps(report))
finally:
    collection.close()
